vision/vae_mnist/vae_mnist.jl

```diff
@@ -0,0 +1,173 @@
+# Variational Autoencoder(VAE)
+#
+# Auto-Encoding Variational Bayes
+# Diederik P Kingma, Max Welling
+# https://arxiv.org/abs/1312.6114
+
+using Base.Iterators: partition
+using BSON
+using CUDAapi: has_cuda_gpu
+using DrWatson: struct2dict
+using Flux
+using Flux: binarycrossentropy, chunk
+using Flux.Data: DataLoader
+using Images
+using Logging: with_logger
+using MLDatasets
+using Parameters: @with_kw
+using ProgressMeter: Progress, next!
+using TensorBoardLogger: TBLogger, tb_overwrite
+using Random
+
+# load MNIST images and return loader
+function get_data(batch_size)
+    xtrain, ytrain = MLDatasets.MNIST.traindata(Float32)
+    # MLDatasets uses HWCN format, Flux works with WHCN 
+    xtrain = reshape(permutedims(xtrain, (2, 1, 3)), 28^2, :)
+    DataLoader(xtrain, ytrain, batchsize=batch_size, shuffle=true)
+end
+
+struct Encoder
+    linear
+    μ
+    logσ
+    Encoder(input_dim, latent_dim, hidden_dim, device) = new(
+        Dense(input_dim, hidden_dim, tanh) |> device,   # linear
+        Dense(hidden_dim, latent_dim) |> device,        # μ
+        Dense(hidden_dim, latent_dim) |> device,        # logσ
+    )
+end
+
+function (encoder::Encoder)(x)
+    h = encoder.linear(x)
+    encoder.μ(h), encoder.logσ(h)
+end
+
+Decoder(input_dim, latent_dim, hidden_dim, device) = Chain(
+    Dense(latent_dim, hidden_dim, tanh),
+    Dense(hidden_dim, input_dim, sigmoid)
+) |> device
+
+function reconstuct(encoder, decoder, x, device)
+    μ, logσ = encoder(x)
+    z = μ + device(randn(Float32, size(logσ))) .* exp.(logσ)
+    μ, logσ, decoder(z)
+end
+
+function model_loss(encoder, decoder, λ, x, device)
+    μ, logσ, decoder_z = reconstuct(encoder, decoder, x, device)
+    len = size(x)[end]
+    # KL-divergence
+    kl_q_p = 0.5f0 * sum(@. (exp(2f0 * logσ) + μ^2 -1f0 - 2f0 * logσ)) / len
+
+    logp_x_z = -sum(binarycrossentropy.(decoder_z, x)) / len
```

Comment:
logitbinarycrossentropy here?

Reply:
sorry, never mind, you have a simgoid as the last activation. If you experience some  nnumerical instabilities, you could try with removing the sigmoid and using logitbinarycorssentropy instead

Reply:
I switched sigmoid + bce to logitbce and inserted sigmoid to `convert_to_image`.